bgzf.c

```diff
@@ -150,6 +168,84 @@ struct __bgzidx_t
     uint64_t ublock_addr;   // offset of the current block (uncompressed data)
 };
 
+/*
+ * Buffers up arguments to hts_idx_push for later use, once we've written all bar
+ * this block.  This is necessary when multiple blocks are in flight (threading)
+ * and fp->block_address isn't known at the time of call as we have in-flight
+ * blocks that haven't yet been compressed.
+ *
+ * NB: this only matters when we're indexing on the fly (writing).
+ * Normal indexing is threaded reads, but we already know block sizes
+ * so it's a simpler process
+ *
+ * Returns 0 on success,
+ *        -1 on failure
+ */
+int bgzf_idx_push(BGZF *fp, hts_idx_t *hidx, int tid, int beg, int end, uint64_t offset, int is_mapped) {
+    hts_idx_cache_entry *e;
+    mtaux_t *mt = fp->mt;
+
+    if (!mt)
+        return hts_idx_push(hidx, tid, beg, end, offset, is_mapped);
+
+    pthread_mutex_lock(&mt->idx_m);
+
+    mt->hts_idx = hidx;
+    hts_idx_cache_t *ic = &mt->idx_cache;
+
+    if (ic->nentries >= ic->mentries) {
+        int new_sz = ic->mentries ? ic->mentries*2 : 1024;
+        if (!(e = realloc(ic->e, new_sz * sizeof(*ic->e)))) {
+            pthread_mutex_unlock(&mt->idx_m);
+            return -1;
+        }
+        ic->e = e;
+        ic->mentries = new_sz;
+    }
+
+    e = &ic->e[ic->nentries++];
+    e->tid = tid;
+    e->beg = beg;
+    e->end = end;
+    e->is_mapped = is_mapped;
+    e->offset = offset & 0xffff;
+    e->block_number = mt->block_number;
+
+    pthread_mutex_unlock(&mt->idx_m);
+
+    return 0;
+}
+
+static int bgzf_idx_flush(BGZF *fp) {
+    mtaux_t *mt = fp->mt;
+
+    if (!mt->idx_cache.e) {
+        mt->block_written++;
+        return 0;
+    }
+
+    pthread_mutex_lock(&mt->idx_m);
+
+    hts_idx_cache_entry *e = mt->idx_cache.e;
+    int i;
+
+    for (i = 0; i < mt->idx_cache.nentries && e[i].block_number == mt->block_written; i++) {
```

Comment:
Is it ever possible that this could get out of synch?  I.e. `e[i].block_number < mt->block_written` which would result in the data staying queued forever...

I'm not sure if it could, but it might be an idea to put in an assert so we get to find out about it if it ever happens.

Reply:
In theory it can't happen as both producer and consumer are incrementing their own variables.  An assert isn't a bad idea though just in case something whacky happens that I didn't think about.